### PR TITLE
1. add ldap data to user object after sync, 2. fix clean_ldap_name regexp

### PR DESCRIPTION
**Pull request by itFrost**
1. add ldap data to user object after sync
feature usage example:

``` py
from django_python3_ldap.auth import LDAPBackend
class MyLDAPBackend(LDAPBackend):
    def authenticate(self, username, password, **kwargs):
        user = super(MyLDAPBackend, self).authenticate(username, password, **kwargs)
		# do something with ldap_data...
        display_name = user.ldap_data['attributes'].get('displayName', (None,))[0]
        return user
```

2. fix clean_ldap_name regexp
Added a dot character

### CODE REVIEW

File: django_python3_ldap/utils.py [hunk 1]
```diff
@@ -17,7 +17,7 @@ def clean_ldap_name(name):
     # values, but I can't find the documentation anywhere!
     # This just removes bad characters entirely, which will
     # work for most cases, but isn't exactly ideal! :O
-    return re.sub("[^a-zA-Z0-9_\-]", "", force_text(name))
+    return re.sub("[^a-zA-Z0-9_\-.]", "", force_text(name))
```
Comment (etianen) [inline django_python3_ldap/utils.py]: Do you have any documentation about allowed characters in LDAP lookups, or some better way of escaping values for use in LDAP lookups?

**Comment by etianen** [comment]
The fix to the clean_ldap_name seems legit, but I'd be curious about where you're getting information about valid LDAP query characters and escaping. I've found nothing, so the regex is very conservative.

I don't think that the approach you've taken for attaching ldap_data to the user object is very clean. I think you can achieve the same thing using a combination of LDAP_AUTH_USER_FIELDS (to map LDAP fields to user model attributes), and LDAP_AUTH_CLEAN_USER_DATA (to modify the LDAP attributes before adding to the user model).

**Comment by itFrost** [comment]
1. About the clean_ldap_name: I think this this page is what we need https://support.microsoft.com/en-us/kb/909264, but I still not sure that now regexp works right.

2. LDAP_AUTH_USER_FIELDS and LDAP_AUTH_CLEAN_USER_DATA is helpful when we fill User model, but not Group.
I can set LDAP_AUTH_USER_FIELDS like this:
LDAP_AUTH_USER_FIELDS = {
    "username": "cn", "email": "mail", "ldap_groups": 'memberOf'
}
Then user_data['ldap_groups'] in LDAP_AUTH_CLEAN_USER_DATA will contain only first group of user.
(because https://github.com/etianen/django-python3-ldap/blob/master/django_python3_ldap/ldap.py#L40)

Even if you add some method to settings for work with ldap attributes, I think that working with ldap_data in auth Backend will be convenient.
I'll try to show it in this example:

from django_python3_ldap.auth import LDAPBackend
from django.contrib.auth.models import Group

class CustomLDAPBackend(LDAPBackend):
    GROUP_PREFIX = 'LDAP_'

    def authenticate(self, username, password,
                     update_credentials=False, update_groups=True, **kwargs):
        user = super(CustomLDAPBackend, self).authenticate(username, password, **kwargs)
        if user:
            if update_groups:
                CustomLDAPBackend.update_groups(user)
        return user

    @staticmethod
    def update_groups(user):
        if user:
            ldap_group_names = user.ldap_data['attributes'].get('memberOf', [])
            group_names = set(CustomLDAPBackend.GROUP_PREFIX + gn.split(',')[0].split('=')[1]
                              for gn in ldap_group_names)

            for group_name in group_names:
                group, created = Group.objects.get_or_create(name=group_name)
                user.groups.add(group)
            for group in user.groups.filter(
                    name__startswith=CustomLDAPBackend.GROUP_PREFIX).exclude(name__in=group_names):
                user.groups.remove(group)

**Comment by etianen** [comment]
Rather than a custom auth backend and sticking the LDAP data on the user model, it feels like the ability to specify a custom LDAP backend might be better. That way, you can override the get_or_create_user method to your heart’s content.

**Comment by mnach** [comment]
Dear, etianen!

As I can see, for now in ldap.py module we don't have a way to specify custom LDAP backend to resolve our needs. All logic have been performed by the instance of Connection class, which have been created by connection function. So, we would like to have a better way than just copy-paste functions: authenticate and connection. 

So, my opinion is that for now we have two solutions:
* Extend flexibility of this application, so that other developers can have a way to collect data from LDAP-object in their models
* Implement functionality of collecting LDAP groups in django Groups - the heart of django-permission functionality.
What would you choose?

We would like to see such changes in this project rather than work with our own fork or use some ugly monkey-patch.

**Comment by etianen** [comment]
Of course. This project is the place to put such logic.

I think the solution is:

- Change the LDAP logic to return all fields.
- A new setting, LDAP_AUTH_SYNC_USER_RELATIONS that points to a function that is run after the user object has been retrieved, with the following signature:

def sync_user_relations(user, ldap_data):
	return user_data

That makes a relatively clean separation, whereby:

LDAP_AUTH_USER_FIELDS and LDAP_AUTH_CLEAN_USER_DATA are for user attributes.

And:

LDAP_AUTH_SYNC_USER_RELATIONS is for related models.

Of course, LDAP_AUTH_SYNC_USER_RELATIONS will be given the user_data["attributes”] dict with all the original lists intact, allowing it to handle multiple relations.

My slight issue with this is that there are now two ways to apply LDAP data to the user. The best way in most circumstances is LDAP_AUTH_USER_FIELDS and LDAP_AUTH_CLEAN_USER_DATA, since it won’t result in any additional database queries. The LDAP_AUTH_SYNC_USER_RELATIONS is needed for any non-user model attributes.

**Comment by mnach** [comment]
I found a manual with characters that need to be escaped in LDAP fields
http://www.rlmueller.net/CharactersEscaped.htm
Seams very legit to me. May be we can use it in clean_ldap_name function?

I think that we can escape those symbols instead of replacing them with empty string. How do you think?

**Comment by etianen** [comment]
Looks legit. I'm not entirely clear on how they suggest escaping non-ascii characters, however. They seem to suggest some sort of hex encoding, but don't mention character sets.

```
You can escape any characters, including foreign and other non-keyboard characters. For example, the following characters.

á é í ó ú ñ

would be escaped as follows (in order):

\E1 \E9 \ED \F3 \FA \F1
```

**Comment by mnach** [comment]
looks like it's extended ascii characters

````python
>>> [c.decode('latin-1') for c in (b'\xe1', b'\xe9', b'\xed', b'\xf3', b'\xfa', b'\xf1')]
['á', 'é', 'í', 'ó', 'ú', 'ñ']
`````
Unfortunately I don't have ldap server with admin permissions, so I can't check it by myself.

**Comment by etianen** [comment]
Right, I've implemented this here:

bed01c7317e68e1efa20f036bc8e84af0777dd7c

I'm being pretty conservative with what isn't escaped, but the way I read that document, escaping too much is better than escaping not enough.

**Pull request closed without merging**